p2p/transport/quicreuse/listener.go

```diff
@@ -0,0 +1,218 @@
+}
+
+type connListener struct {
+	l       quic.Listener
+	conn    pConn
+	running chan struct{}
+	addrs   []ma.Multiaddr
+
+	mx        sync.Mutex
+	protocols map[string]protoConf
+}
+
+func newConnListener(c pConn, quicConfig *quic.Config, enableDraft29 bool) (*connListener, error) {
+	localMultiaddrs := make([]ma.Multiaddr, 0, 2)
+	a, err := ToQuicMultiaddr(c.LocalAddr(), quic.Version1)
+	if err != nil {
+		return nil, err
+	}
+	localMultiaddrs = append(localMultiaddrs, a)
+	if enableDraft29 {
+		a, err := ToQuicMultiaddr(c.LocalAddr(), quic.VersionDraft29)
+		if err != nil {
+			return nil, err
+		}
+		localMultiaddrs = append(localMultiaddrs, a)
+	}
+	cl := &connListener{
+		protocols: map[string]protoConf{},
+		running:   make(chan struct{}),
+		conn:      c,
+		addrs:     localMultiaddrs,
+	}
+	tlsConf := &tls.Config{
+		GetConfigForClient: func(info *tls.ClientHelloInfo) (*tls.Config, error) {
+			cl.mx.Lock()
+			defer cl.mx.Unlock()
+			for _, proto := range info.SupportedProtos {
+				if entry, ok := cl.protocols[proto]; ok {
+					conf := entry.tlsConf
+					if conf.GetConfigForClient != nil {
+						return conf.GetConfigForClient(info)
+					}
+					return conf, nil
+				}
+			}
+			return nil, fmt.Errorf("no supported protocol found. offered: %+v", info.SupportedProtos)
+		},
+	}
+	quicConf := quicConfig.Clone()
+	quicConf.AllowConnectionWindowIncrease = cl.allowWindowIncrease
+	ln, err := quicListen(c, tlsConf, quicConf)
+	if err != nil {
+		return nil, err
+	}
+	cl.l = ln
+	go cl.Run() // This go routine shuts down once the underlying quic.Listener is closed (or returns an error).
+	return cl, nil
+}
+
+func (l *connListener) allowWindowIncrease(conn quic.Connection, delta uint64) bool {
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	conf, ok := l.protocols[conn.ConnectionState().TLS.ConnectionState.NegotiatedProtocol]
+	if !ok {
+		return false
+	}
+	return conf.allowWindowIncrease(conn, delta)
+}
+
+func (l *connListener) Add(tlsConf *tls.Config, allowWindowIncrease func(conn quic.Connection, delta uint64) bool, onRemove func()) (Listener, error) {
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	if len(tlsConf.NextProtos) == 0 {
+		return nil, errors.New("no ALPN found in tls.Config")
+	}
+
+	for _, proto := range tlsConf.NextProtos {
+		if _, ok := l.protocols[proto]; ok {
+			return nil, fmt.Errorf("already listening for protocol %s", proto)
+		}
+	}
+
+	ln := newSingleListener(l.l.Addr(), l.addrs, func() {
+		l.mx.Lock()
+		for _, proto := range tlsConf.NextProtos {
+			delete(l.protocols, proto)
+		}
+		l.mx.Unlock()
+		onRemove()
+	})
+	for _, proto := range tlsConf.NextProtos {
+		l.protocols[proto] = protoConf{
+			ln:                  ln,
+			tlsConf:             tlsConf,
+			allowWindowIncrease: allowWindowIncrease,
+		}
+	}
+	return ln, nil
+}
+
+func (l *connListener) Run() error {
+	defer close(l.running)
+	defer l.conn.DecreaseCount()
+	for {
+		conn, err := l.l.Accept(context.Background())
+		if err != nil {
+			return err
```

Comment:
We could at least bubble up the error to the listener's `Accept` call (since this is what this goroutine is for) and then let the caller close the listener (and thus the remove handlers) later. Something like this: https://gist.github.com/MarcoPolo/dd25fb55444bd8a22f801257061c74f1

Reply:
Implemented this.